iavl/diff.py

```diff
@@ -84,102 +70,130 @@ def successor_prune(n: PersistedNode) -> (bool, bool):
             if n.version <= version:
                 shared.add(n.hash)
             elif n.is_leaf():
-                new.append(n)
+                new.append(KVPair(key=n.key, value=n.value))
+                new_keys.add(n.key)
 
     def prune(n: PersistedNode) -> (bool, bool):
         b = n.hash in shared
         return b, b
 
     if root:
-        orphaned = [
-            n
+        deleted = [
+            KVPair(delete=True, key=n.key)
             for n in visit_iavl_nodes(get_node, prune, root)
-            if n.is_leaf() and n.hash not in shared
+            if n.is_leaf() and n.hash not in shared and n.key not in new_keys
         ]
     else:
-        orphaned = []
+        deleted = []
 
-    return split_operations(orphaned, new)
+    changeset = new + deleted
+    changeset.sort(key=lambda n: n.key)
+    return changeset
 
 
 def apply_change_set(tree: Tree, changeset: ChangeSet):
     """
     changeset: the result of `state_changes`
     """
-    for key, op, arg in changeset:
-        if op == Op.Insert:
-            tree.set(key, arg)
-        elif op == Op.Update:
-            _, value = arg
-            tree.set(key, value)
-        elif op == Op.Delete:
-            tree.remove(key)
+    for pair in changeset:
+        if pair.delete:
+            tree.remove(pair.key)
         else:
-            raise NotImplementedError(f"unknown op {op}")
+            tree.set(pair.key, pair.value)
 
 
-class StoreKVPairs(ProtoEntity):
+def append_change_set(fp, version: int, changeset: ChangeSet):
     """
-    protobuf format compatible with file streamer output
-    store an additional original value, it's empty for insert operation.
-    """
-
-    # the store key for the KVStore this pair originates from
-    store_key = Field("string", 1)
-    # true indicates a delete operation
-    delete = Field("bool", 2)
-    key = Field("bytes", 3)
-    value = Field("bytes", 4)
-    original = Field("bytes", 5)
+    write change set to file, file format:
 
-    def as_json(self):
-        d = {"key": binascii.hexlify(self.key).decode()}
-        if self.store_key:
-            d["store_key"] = self.store_key
-        if self.value:
-            d["value"] = binascii.hexlify(self.value).decode()
-        if self.original:
-            d["original"] = binascii.hexlify(self.original).decode()
-        if self.delete:
-            d["delete"] = True
-        return d
-
-
-def write_change_set(fp, changeset: ChangeSet, store=""):
-    """
-    write change set to file, compatible with the file streamer output.
+    ```
+    version: varint
+    size: varint # the total size of kv-pairs, so we can skip faster
+    kv-pairs: length prefixed proto msg
```

Comment:
not sure if we need delimiter for this len, msg, len, msg,…

Reply:
do you mean: `msg, msg, ...`, remove the length prefix?

Reply:
I think protobuf won't work an end mark, unless we encode the static fields directly.

Reply:
I mean align like this [delimiter](https://github.com/cosmos/gogoproto/blob/main/io/varint.go#L45)

Reply:
what's the difference?